xds/internal/resolver/cluster_specifier_plugin_test.go

```diff
@@ -20,349 +20,466 @@ package resolver
 
 import (
 	"context"
+	"encoding/json"
+	"fmt"
 	"testing"
 
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes"
 	"github.com/google/go-cmp/cmp"
+	"github.com/google/uuid"
 	"google.golang.org/grpc/balancer"
 	"google.golang.org/grpc/internal"
+	"google.golang.org/grpc/internal/envconfig"
 	iresolver "google.golang.org/grpc/internal/resolver"
+	"google.golang.org/grpc/internal/testutils"
+	xdsbootstrap "google.golang.org/grpc/internal/testutils/xds/bootstrap"
+	"google.golang.org/grpc/internal/testutils/xds/e2e"
 	"google.golang.org/grpc/resolver"
 	"google.golang.org/grpc/serviceconfig"
 	"google.golang.org/grpc/xds/internal/balancer/clustermanager"
 	"google.golang.org/grpc/xds/internal/clusterspecifier"
-	"google.golang.org/grpc/xds/internal/xdsclient/xdsresource"
+	"google.golang.org/protobuf/types/known/anypb"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+
+	v3listenerpb "github.com/envoyproxy/go-control-plane/envoy/config/listener/v3"
+	v3routepb "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
 )
 
 func init() {
-	balancer.Register(cspB{})
+	balancer.Register(cspBalancerBuilder{})
+	clusterspecifier.Register(testClusterSpecifierPlugin{})
 }
 
-type cspB struct{}
+// cspBalancerBuilder is a no-op LB policy which is referenced by the
+// testClusterSpecifierPlugin.
+type cspBalancerBuilder struct{}
 
-func (cspB) Build(cc balancer.ClientConn, opts balancer.BuildOptions) balancer.Balancer {
+func (cspBalancerBuilder) Build(cc balancer.ClientConn, opts balancer.BuildOptions) balancer.Balancer {
 	return nil
 }
 
-func (cspB) Name() string {
+func (cspBalancerBuilder) Name() string {
 	return "csp_experimental"
 }
 
-type cspConfig struct {
+type cspBalancerConfig struct {
+	serviceconfig.LoadBalancingConfig
 	ArbitraryField string `json:"arbitrary_field"`
 }
 
-// TestXDSResolverClusterSpecifierPlugin tests that cluster specifier plugins
-// produce the correct service config, and that the config selector routes to a
-// cluster specifier plugin supported by this service config (i.e. prefixed with
-// a cluster specifier plugin prefix).
-func (s) TestXDSResolverClusterSpecifierPlugin(t *testing.T) {
-	xdsR, xdsC, tcc, cancel := testSetup(t, setupOpts{target: target})
-	defer xdsR.Close()
-	defer cancel()
+func (cspBalancerBuilder) ParseConfig(lbCfg json.RawMessage) (serviceconfig.LoadBalancingConfig, error) {
+	cfg := &cspBalancerConfig{}
+	if err := json.Unmarshal(lbCfg, cfg); err != nil {
+		return nil, err
+	}
+	return cfg, nil
+
+}
+
+// testClusterSpecifierPlugin is a test cluster specifier plugin which returns
+// an LB policy configuration specifying the cspBalancer.
+type testClusterSpecifierPlugin struct {
+}
```

Comment:
Why did we need to create a new plugin? Did my old tests not use a new plugin?

Reply:
The old tests didn't need a plugin implementation since they could pass the appropriate plugin configuration in the internal update data structures. But since the new tests rely on sending actual xDS resources (and thereby test thing from the time of receiving the resource to the time the RPC is made), we need a real plugin implementation for the one specified in the resource.